### DISCUSSION TITLE
Adding various types of diagrams (such as a circular diagram) to a node in Freeplane.

### DISCUSSION THREAD

**Discussion by GorBinance** [freeplane/freeplane#1115] [Ideas]
Dear friends, I suggest adding a diagram function to the nodes with a feature of a numbering marker to calculate the percentage of the diagrams. I have attached a visual example of the implementation as well as a saved mind map.
[Doing  - balance wheel.zip](https://github.com/freeplane/freeplane/files/11339363/Doing.-.balance.wheel.zip)
![Explanation](https://user-images.githubusercontent.com/114777637/234754283-ee347985-3aee-4de1-947d-e9a1f9d7bd51.jpg)
![Sketch](https://user-images.githubusercontent.com/114777637/234754294-740d6ff3-9ffb-420b-b0fd-cab91f293428.jpg)
[Doing - Balance wheel.xlsx](https://github.com/freeplane/freeplane/files/11339374/Doing.-.Balance.wheel.xlsx)

**Comment by GorBinance** [comment]
It is also possible to establish dependencies between the parent node and the child nodes, as well as all subsequent nodes. If some numbering markers are empty, they will be considered as 0.
![node](https://user-images.githubusercontent.com/114777637/234758781-10e0d3c4-e8f3-45cb-b9e8-fa77f8604fba.jpg)

**Reply by GorBinance** [reply]
It is also possible to establish dependencies between the parent node and the child nodes, as well as all subsequent nodes. If some numbering markers are empty, they will be considered as 0.

This aspect with the "numbering marker" fits very well into the already implemented attributes and at the same time expands the possibilities and simplifies the implementation and visualization of diagrams.
![atributs](https://user-images.githubusercontent.com/114777637/234790656-dcc504aa-0855-4cca-92e7-002de8908bd6.jpg)

**Reply by EdoFro** [reply]
hi @GorBinance,
I think you are searching for [Formulas in Freeplane](https://docs.freeplane.org/scripting/Formulas.html).

Formulas can be used in node's text, details, attributes and/or notes.

hope this helps,
edo

**Comment by EdoFro** [comment]
@GorBinance,
here is also an example on how you can add a graph to Freeplane using the [jFreechart lib](https://www.jfree.org/jfreechart/).

In this example it takes the data from the 'mm' attributes in the child nodes from the "Rainfall" node and builds a graph with it:
![image](https://user-images.githubusercontent.com/35700575/235252273-3ab71cf5-2067-44dc-9bf7-0760fb0beb18.png)

Here is the mindmap:
[BarGraph example.zip](https://github.com/freeplane/freeplane/files/11357287/BarGraph.example.zip)

the script code is:

```groovy
@Grab(group='org.jfree', module='jfreechart', version='1.5.4')

import groovy.swing.SwingBuilder
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartFrame
import org.jfree.chart.plot.PlotOrientation
import org.jfree.data.category.DefaultCategoryDataset

def createChart(data) {
    def dataset = new DefaultCategoryDataset()
    data.each { month, rain -> dataset.addValue(rain, "Rainfall", month) }
    ChartFactory.createBarChart(
            "Rainfall per Month",
            "Month",
            "Rainfall (mm)",
            dataset,
            PlotOrientation.VERTICAL,
            true,
            true,
            false
    )
}

def showChart(data) {
    def chart = createChart(data)
    def frame = new ChartFrame("Rainfall per Month", chart)
    frame.pack()
    frame.setVisible(true)
}

def createForm() {
    def builder = new SwingBuilder()
        def frame = builder.frame(title: "Rainfall Graph", size: [400, 400]) {
            panel() {
                button('Show Graph', actionPerformed: {
                    def data = getData()
                    showChart(data)
                })
            }
        }
        frame.show()
}

def getData(){
    def attrName = 'mm'
    def nodo = getNode()
    def data =[:]
    nodo.children.each{n ->
        data[n.text] = n[attrName].to.num0
    }
    return data
}

def getNode(){
    node.mindMap.node('ID_1824238975')
}

c.select(getNode())

createForm()

```

Hope it helps

**Reply by EdoFro** [reply]
you can use chatGPT to modify it, asking for a pie chart for example.

**Reply by EdoFro** [reply]
Just for the record:
I created this example a few days ago just to see if chatGPT could really code. I asked it to code an example program in Groovy and in Kotlin:
![image](https://user-images.githubusercontent.com/35700575/235272205-0e4d95d9-4073-4b53-a3f1-14f175f4942a.png)

then I took the code in Groovy and tested it as a script in Freeplane. I had to do very few modifications to it to run.

-----

Today I modified it, so it takes the input data from the mind map, and put it here as an example, but all the jFreeChart thing was made by chatGPT.

**Comment by dpolivaev** [comment]
I am considering the integration of JFreeChart into Freeplane.
For each node, it could create a chart using a given attribute name and collect the information from all child nodes or descendant nodes. Charts could be embedded into the map and exported as svg files.

How do you think it is interesting for casual users? 
Advanced folks can certainly do it using scripts as shown by @EdoFro .

**Reply by EdoFro** [reply]
I think it's a great idea.
A first step would it be to be able to embed the chart into the `node.externalObject` through a script,

Maybe a new API method like `node.externalObject.setChart( org.jfree.chart.JFreeChart )`

having that, we can play a little and discover what can be done and think on a better user interface for casual users that want to use charts without scripting.

Or maybe through formulas?

jm2c,
edo

**Reply by Illegit** [reply]
Nice to have this tool, +1

**Reply by bepolymathe** [reply]
Yes me too +1 !

**Comment by EdoFro** [comment]
Another idea:

using formulas
if the result is a JFreeChart Object then it displays it in the node:

something like:
```groovy
=    
def dataset = new DefaultCategoryDataset()

node.children.each{n ->
   dataset.addValue(n['mm'].to.num0, "Rainfall", n.text)
}

def myChart = ChartFactory.createBarChart(
                    "Rainfall per Month",
                    "Month",
                    "Rainfall (mm)",
                    dataset,
                    PlotOrientation.VERTICAL,
                    true,
                    true,
                    false
            )

return myChart
```

**Comment by Illegit** [comment]
Just to ask for help: I was using the suggestion of EdoFro and  I share this map of mine with the results, where I have a question about how to create multiple lines and multiple bars diagram with Jfreechart, I made all the others you see in the map but wasn't able to solve the multiple ones, any suggestions? ChatGPT was always wrong in something... and I didn't find the "quid" to solve them.

Thanks in advance and maybe my map can be helpful for somebody.
[Formulas and Graphs.zip](https://github.com/freeplane/freeplane/files/11359772/Formulas.and.Graphs.zip)

**Comment by Illegit** [comment]
here are the solutions to my above questions. Anyway is welcome the dimitry point is much more effective than my zip file.
[Uploading Formulas and Graphs.zip…]()

**Comment by Illegit** [comment]
here the zip
[Formulas and Graphs.zip](https://github.com/freeplane/freeplane/files/11362265/Formulas.and.Graphs.zip)

**Reply by EdoFro** [reply]
Hi @Illegit,
I had some problems with the scripts. Maybe you have an older JFreeChart library version?

I changed them to work in my computer.
I also removed the first frame with the button, so now it displays the chart immediately.

[Formulas_and _Graphs_RevEdo.zip](https://github.com/freeplane/freeplane/files/11367455/Formulas_and._Graphs_RevEdo.zip)

Great job. Thanks,

edo

**Reply by Illegit** [reply]
Thank EdoFro:
I use the JFreeChart 1.5.4, that was collected years ago in this forum, all the script are running with it. But to enjoy of your removing the button may you provide your version maybe in zip file of the JFreeChart, with my version of JFreeChart only your simpler file are working the others not.

**Reply by GorBinance** [reply]
Dear friends, could you please advise if there is a full-fledged API documentation in Freeplane? I've been exploring Freeplane for about a week and a half now and I'm already in love with it, but I'm having some trouble with scripts, even with the help of Chat GPT and Bing (by the way, Bing is better at writing scripts). I would greatly appreciate any advice. The real problem is navigating the map through scripts, i.e., interacting with parent and child nodes, as well as searching for nodes based on various indicators and values, such as ID or name.

**Reply by EdoFro** [reply]
menu Help->Freeplane API…

**Comment by Illegit** [comment]
Hi EdoFro, this JFreeCHart topic is not clear to me I found in my PC the version 1.5.4 (I saw it available in GITHUB)
but i discovered that my freeplane is using the 1.0.13, I have updated to 1.5.4 changed the jar and everything is OK. Thankyou, your version is working nicely.

**Comment by GorBinance** [comment]
Dear friends and like-minded individuals, for the past 2 days I have been working in collaboration with Bing, ChatGPT 3.5, and ChatGPT 4 to script an adaptive program that can be controlled by GPT-like language models. I am creating something like a balance wheel and organizer under the control of a language model, so that it can press buttons when I make a request, much like taking a test on how much time I spent on a particular aspect of my balance wheel. The language model can also control the map. Initially, I want to have buttons in a manual format, like a dashboard, where I can answer questions at the end of the day about how much time I spent on each aspect. However, I have trouble finding nodes, so to speak, node cretinism XD is a problem for me. I even considered creating a block programming system like in Zennoposter, which could make the scripts more convenient, secure, and reliable. And of course, it would be more user-friendly to interact with Freeplane as I attached a rough sketch of the map.
[ohhhhh.zip](https://github.com/freeplane/freeplane/files/11378705/ohhhhh.zip)

**Reply by GorBinance** [reply]
This issue is solved using def targetNode=node.mindMap.node('ID_....') node search by ID.
